### PR TITLE
[ShaderGraph] Reimplement redirect nodes using GTF

### PR DESCRIPTION
**Pull request by dhsavell**
### Purpose of this PR

This PR ports redirect nodes from the current version of ShaderGraph with best possible parity. Here is a quick overview with current SG on the left and our implementation on the right. See some demos below.

NB: This PR does not implement any new control point features.

---
### Testing status

Because we are planning on tests later, most manual testing was done for parity. Lots of gifs ahead:

| Current | New | Notes |
| --- | --- | --- |
| image | image | Appearance |
| image | image | Validation (see known issues) |
| kodrzaCAa8 | rlkw72IZDI | Adding Redirects |
| eyrtdEkak8 | 9YLLbJ8fhZ | Deleting from keyboard |
| tRnvHkvCA3 | wGw4ddHVDA | Deleting from context menu (see known issues) |
| EnCmqJpzXD | HNULCIz74l | Port Type |
| P7Jjg8glx1 | zmKeRhBqbe | Copy & Paste |

Known issues & caveats:

- Pasting a redirect whose source node has been deleted causes a NullReferenceException from GTF (reported: [GTF-445](https://jira.unity3d.com/browse/GTF-445)).
- Edges do not disappear immediately when deleting nodes from context menu (known: [GTF-438](https://jira.unity3d.com/browse/GTF-438)).
- Deleting an edge in the tool doesn't delete it from the graph data yet (for any nodes, not just redirects). Creating edges should work.
- Badges are being refactored so we are using the default warning appearance for now, despite differing from the original.

---
### Comments to reviewers

Some of the port compatibility and edge creation code was refactored because it did not yet implement connections from data-backed nodes to GTF-only ones.

Please do let me know if class/member names don't make sense. Some of them had quite a few iterations since their initial implementations.

### CODE REVIEW

File: com.unity.shadergraph/Editor/GraphUI/GraphElements/ModelUI/RedirectNode.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            PartList.AppendPart(InOutPortContainerPart.Create(portContainerPartName, Model, this, ussClassName));
+            AddToClassList("sg-redirect-node");
+            this.AddStylesheet("RedirectNode.uss");
+        }
+
+        protected override void BuildContextualMenu(ContextualMenuPopulateEvent evt)
+        {
+            base.BuildContextualMenu(evt);
+
+            // TODO: Remove
```
Comment (jessebarker) [inline com.unity.shadergraph/Editor/GraphUI/GraphElements/ModelUI/RedirectNode.cs:24]: What does this mean?  Remove what? And why?  (just wary of code going in with todo notes that someone other than the author might not understand)

**Review by jessebarker — COMMENTED** [review]
LGTM.

**Review by jessebarker — APPROVED** [review]

File: com.unity.shadergraph/Editor/GraphUI/GraphElements/ModelUI/RedirectNode.cs (hunk 1 above)
Comment (dhsavell) [reply com.unity.shadergraph/Editor/GraphUI/GraphElements/ModelUI/RedirectNode.cs:24]: Oops, I didn't finish writing my sentence. It was for a temporary context menu item below. But on second look, it's probably not useful to check into staging -- I'll go ahead and remove it.

**Review by esmelusina — APPROVED** [review]
This looks good. Usage of CSO and traversal to discover and resolve connections is a well thought out approach. This also looks like it will handle disruptions to type representation as well- props! Looks great.

**Pull request merged**